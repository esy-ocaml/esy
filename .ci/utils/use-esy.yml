# steps to install esy globally

steps:
- script: "yarn global add @esy-nightly/esy@0.6.8-bbdfb8"
  displayName: "Install esy (via yarn)"
  condition: ne(variables['AGENT.OS'], 'Windows_NT')
  
- script: |
    yarn config set prefix C:\
    yarn global add @esy-nightly/esy@0.6.8-bbdfb8
  displayName: "Install esy (via yarn)"
  condition: eq(variables['AGENT.OS'], 'Windows_NT')

- bash: |
    YARN_BIN_DIR=$(yarn global bin)
    echo $YARN_BIN_DIR
    echo "##vso[task.setvariable variable=YARN_BIN_DIR]$YARN_BIN_DIR"
  displayName: "Setting $YARN_BIN_DIR"

- bash: |
    echo "##vso[task.setvariable variable=PATH]$PATH:$YARN_BIN_DIR" 
  displayName: "Ensure yarn install global tools are available on $PATH"
  condition: ne(variables['AGENT.OS'], 'Windows_NT')

- powershell: Write-Host "##vso[task.setvariable variable=PATH;]${env:PATH};${env:YARN_BIN_DIR}"
  displayName: "Ensure yarn install global tools are available on $PATH"
  condition: eq(variables['AGENT.OS'], 'Windows_NT')
