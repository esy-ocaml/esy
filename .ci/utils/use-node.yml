# steps to use node on agent

steps:
  - task: NodeTool@0
    displayName: "Use Node 15.x"
    inputs:
      versionSpec: 15.x
