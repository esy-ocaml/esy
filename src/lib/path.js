/**
 * @flow
 */

import invariant from 'invariant';
const path = require('path');

/**
 * Absolute path.
 */
export opaque type AbsolutePath: ConcretePath = string;

/**
 * Path with no `.` or `..` or any bash variables in it.
 */
export opaque type ConcretePath: Path = string;

/**
 * Path which can contain bash/make variables in it, cannot be used in
 * filesystem operations.
 */
export opaque type AbstractPath: Path = string;

export opaque type Path: string = AbsolutePath | ConcretePath | AbstractPath;

export function absolute(p: string): AbsolutePath {
  p = sanitizeAbsolutePath(p);
  return p;
}

export function concrete(p: string): ConcretePath {
  p = sanitizeConcretePath(p);
  return p;
}

export function abstract(p: string): AbstractPath {
  return p;
}

export function join<B: Path | string>(base: B, ...p: string[]): B {
  return (path.join(base, ...p): any);
}

export function basename<P: Path | string>(p: P): string {
  return path.basename(p);
}

export function dirname<P: Path | string>(p: P): P {
  return (path.dirname(p): any);
}

export function length(p: ConcretePath): number {
  return p.length;
}

export function toString(p: Path | string): string {
  return p;
}

export const sep: typeof path.sep = path.sep;

export const relative: typeof path.relative = path.relative;

function sanitizeAbsolutePath(p) {
  p = sanitizeConcretePath(p);
  invariant(path.isAbsolute(p), 'Should be absolute path but got: %s', p);
  return p;
}

function sanitizeConcretePath(p) {
  // NOTE: regexpes are stateful, this is why we are creating them here.
  // Consider resetting their state after matching instead.
  const REMOVE_TRAILING_SLASH_RE = /\/+$/g;
  const ENVIRONMENT_VAR_RE = /\$[a-zA-Z_]/g;
  const DOT_PATH_SEGMENT_RE = /\/\.\//g;
  const DOT_DOT_PATH_SEGMENT_RE = /\/\.\.\//g;
  const SANITIZE_SLASH_RE = /\/+/g;

  invariant(
    DOT_PATH_SEGMENT_RE.exec(p) == null,
    'Should not contain "." segments: %s',
    p,
  );
  invariant(
    DOT_DOT_PATH_SEGMENT_RE.exec(p) == null,
    'Should not contain ".." segments: %s',
    p,
  );
  invariant(
    ENVIRONMENT_VAR_RE.exec(p) == null,
    'Path should not contain variable references: %s',
    p,
  );
  p = path.normalize(p);
  p = p.replace(REMOVE_TRAILING_SLASH_RE, '');
  p = p === '' ? '/' : p;
  return p;
}
