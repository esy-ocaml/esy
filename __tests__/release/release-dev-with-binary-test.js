/**
 * @flow
 */

import * as path from 'path';
import {defineTestCaseWithShell} from '../utils';

defineTestCaseWithShell(
  path.join(__dirname, 'fixtures', 'with-binary'),
  `
    run esy release dev
    run cd _release/dev

    run npmGlobal pack
    run npmGlobal -g install ./with-binary-0.1.0.tgz

    assertStdout say-hello.exe HELLO
  `,
);
