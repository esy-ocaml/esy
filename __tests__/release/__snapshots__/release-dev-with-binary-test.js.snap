// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`installation 1`] = `
"<root>
| bin
| | say-hello.exe
| | with-binary-esy-sandbox
| etc
| lib
| | node_modules
| | | with-binary
| | | | _esy
| | | | | bin
| | | | | | esy
| | | | | etc
| | | | | lib
| | | | | | node_modules
| | | | | | | esy
| | | | | | | | _esyInstallCache-3.x.x
| | | | | | | | | v1
| | | | | | | | | | .tmp
| | | | | | | | bin
| | | | | | | | | esy
| | | | | | | | | esy.js
| | | | | | | | node_modules
| | | | | | | | | .bin
| | | | | | | | | | esy-install
| | | | | | | | | @esy-ocaml
| | | | | | | | | | esy-install
| | | | | | | | | | | bin
| | | | | | | | | | | | esy-install
| | | | | | | | | | | | esy-install.js
| | | | | | | | | | | package.json
| | | | | | | | | | esy-opam
| | | | | | | | | | | dist
| | | | | | | | | | | | esy-opam.js
| | | | | | | | | | | | esy-opam.js.flow
| | | | | | | | | | | package.json
| | | | | | | | | | | README.md
| | | | | | | | package.json
| | | | .bin
| | | | | say-hello.exe
| | | | | with-binary-esy-sandbox
| | | | package.json
| | | | postinstall.sh
| | | | prerelease.sh
| | | | r
| | | | | _esyEjectRoot
| | | | | | bin
| | | | | | | fastreplacestring.cpp
| | | | | | | fastreplacestring.exe
| | | | | | | get-store-path
| | | | | | | realpath
| | | | | | | realpath.c
| | | | | | | render-env
| | | | | | | runtime.sh
| | | | | | command-env
| | | | | | eject-env
| | | | | | Makefile
| | | | | | records
| | | | | | | final-install-path-set.txt
| | | | | | | final-install-path-set.txt.in
| | | | | | | store-path.txt
| | | | | | | store-path.txt.in
| | | | | | sandbox.sb.in
| | | | | 3______________________________
| | | | | | i
| | | | | | | with_binary-0.1.0
| | | | | | | | bin
| | | | | | | | | say-hello.exe
| | | | | | | | doc
| | | | | | | | etc
| | | | | | | | lib
| | | | | | | | man
| | | | | | | | sbin
| | | | | | | | share
| | | | | node_modules
| | | | | | .cache
| | | | | | | _esy
| | | | | | | | store
| | | | | | | | | b
| | | | | | | | | i
| | | | | | | | | s
| | | | | package.json
| | | | | say-hello.exe
| | | | | test.js
| | | | records
| | | | | done.txt
| | | | | recordedClientBuildStorePath.txt
| | | | | recordedClientInstallStorePath.txt
| | | | | recordedCoppiedArtifacts.txt
| | | | | recordedServerBuildStorePath.txt"
`;

exports[`installation stdout 1`] = `
"/tmp/TMPDIR/npm/bin/say-hello.exe -> /tmp/TMPDIR/npm/lib/node_modules/with-binary/.bin/say-hello.exe
/tmp/TMPDIR/npm/bin/with-binary-esy-sandbox -> /tmp/TMPDIR/npm/lib/node_modules/with-binary/.bin/with-binary-esy-sandbox

> with-binary@0.1.0 postinstall /tmp/TMPDIR/npm/lib/node_modules/with-binary
> ./postinstall.sh

*** Installing esy for the release...
*** Installing dependencies...
error Could not open cafile: EISDIR: illegal operation on a directory, read
*** Ejecting build environment...
*** Building the release...
[1;37m*** with-binary @ 0.1.0: building from source...[0m
[0;32m*** with-binary @ 0.1.0: build complete[0m
*** Decompressing artefacts...
*** Relocating artefacts to the final destination...
+ with-binary@0.1.0
added 1 package in X.XXXs"
`;

exports[`release 1`] = `
"<root>
| dev
| | .bin
| | | say-hello.exe
| | | with-binary-esy-sandbox
| | package.json
| | postinstall.sh
| | prerelease.sh
| | r
| | | package.json
| | | say-hello.exe
| | | test.js
| | records"
`;

exports[`release stdout 1`] = `
"*** Creating dev-type release for with-binary...
*** Release package created

    Location: _release/dev
    Release Type: dev"
`;
